analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ObjectsShouldNotBeDisposedMoreThanOnceBase.cs

```diff
@@ -0,0 +1,49 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
+
+public abstract class ObjectsShouldNotBeDisposedMoreThanOnceBase : SymbolicRuleCheck
+{
+    protected const string DiagnosticId = "S3966";
+    protected const string MessageFormat = "Refactor this code to make sure '{0}' is disposed only once.";
+
+    private static readonly string[] DisposeMethods = { "Dispose", "DisposeAsync", "Close" };
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        if (context.Operation.Instance.AsInvocation() is { } invocation
+            && DisposeMethods.Contains(invocation.TargetMethod.Name))
```

Comment:
Does this work with implicitly implemented `IDisposable`interface? Let's add UT

Reply:
You mean `explicit` maybe? 
Like:
```csharp
    void IDisposable.Dispose() {  }
```

Reply:
Yeah, explicit... these stupid C# interfaces :/

Reply:
I've seen `DisposableWithExplicitImplementation`, but I don't see it used anywhere

Reply:
Right under but with copy-paste error.

Fixed.